src/server/game/Craft/Crafting.cpp

```diff
@@ -0,0 +1,243 @@
+                {
+                    bool hasItem = std::count_if(std::cbegin(requiredItems), std::cend(requiredItems), [providedReagent](ItemSparseEntry const* itemSparse) {
+                        return itemSparse->ID == uint32(providedReagent.ItemID);
+                    });
+
+                    if (hasItem)
+                        bonusTreeIds.push_back(reagentItem->ItemBonusTreeID);
+
+                    hasOneMatchingItem = hasOneMatchingItem || hasItem;
+                }
+            }
+        }
+
+        // If any item provided do not match, probably a cheater, return
+        if (!hasOneMatchingItem)
+            return SpellCastResult::SPELL_FAILED_NEED_MORE_ITEMS;
+    }
+
+    uint32 skillLevel = GetSkillLevelForCraft();
+    uint32 difficultyPercentMatched = round(float(skillLevel) / float(craftingDifficulty) * 100.f);
+
+    uint32 qualityTier = sDB2Manager.GetCraftingQualityTierByDifficultyPercent(_craftingData->CraftingDifficultyID, difficultyPercentMatched);
+
+    uint32 craftedItemId = GetCraftedItemIdForQuality(qualityTier);
+
+    // If we didn't found anything, we have a problem. Return and don't take reagent from player.
+    if (!craftedItemId)
+        return SpellCastResult::SPELL_FAILED_ERROR;
+
+    ItemContext context = _spell->GetSpellInfo()->HasAttribute(SPELL_ATTR0_IS_TRADESKILL) ? ItemContext::Trade_Skill : ItemContext::NONE;
+
+    std::vector<int32> bonusLists;
+    for (int32 bonusTreeID : bonusTreeIds)
+    {
+        std::vector<int32> treeBonusList = ItemBonusMgr::GetBonusListsForItem(craftedItemId, { context, {}, {}, bonusTreeID, qualityTier });
+        bonusLists.insert(std::end(bonusLists), std::begin(treeBonusList), std::end(treeBonusList));
+    }
+
+    if (Item* createdItem = _spell->DoCreateItem(craftedItemId, context, &bonusLists, false))
+    {
+        InitCraftingStatModifier(createdItem);
+
+        createdItem->AddToWorld();
+        createdItem->SendUpdateToPlayer(_player);
+    }
+
+    // Item has been crafted, remove reagent from player
+    for (auto const& [itemId, itemCount] : usedReagentsToDestroy)
+        _player->DestroyItemCount(itemId, itemCount, true);
+
+    return SpellCastResult::SPELL_CAST_OK;
+}
+
+uint32 Crafting::GetSkillLevelForCraft()
+{
+    uint32 skillId = GetSkillIdForSpell(_spell->GetSpellInfo()->Id);
+    uint32 skillValue = _player->GetSkillValue(skillId);
+
+    uint32 totalReagentWeigth = CalculateTotalReagentWeights();
+
+    CraftingDifficultyEntry const* craftingDifficultyEntry = sCraftingDifficultyStore.LookupEntry(_craftingData->CraftingDifficultyID);
+    if (!craftingDifficultyEntry)
+        return skillValue;
+
+    float maximumReagentBonus = round(float(_craftingData->CraftingDifficulty) * float(craftingDifficultyEntry->CraftSkillBonusPercent) / 100.f);
+
+    for (ModifiedCraftingSpellSlotEntry const* spellSlot : sDB2Manager.GetMCRSpellSlotBySpell(_spell->GetSpellInfo()->Id))
```

Comment:
this call is done multiple times in different functions, you could retrieve it once in DoCraft and pass it as const& to GetSkillLevelForCraft() and CalculateTotalReagentWeights() and save resources